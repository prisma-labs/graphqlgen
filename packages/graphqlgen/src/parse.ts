import * as Ajv from 'ajv'
import * as chalk from 'chalk'
import * as fs from 'fs'
import * as yaml from 'js-yaml'
import * as Path from 'path'
import * as tsNode from 'ts-node'
import { print } from 'graphql'
import { importSchema } from 'graphql-import'
import {
  GraphQLGenDefinition,
  Language,
  Models,
  File,
} from 'graphqlgen-json-schema'
import schema = require('graphqlgen-json-schema/dist/schema.json')

import { ContextDefinition, ModelMap, Model } from './types'
import {
  getAbsoluteFilePath,
  getImportPathRelativeToOutput,
} from './path-helpers'
import { getTypeToFileMapping, replaceAll, normalizeFilePath } from './utils'
import {
  extractTypes,
  extractGraphQLTypesWithoutRootsAndInputsAndEnums,
  GraphQLTypes,
} from './source-helper'
import { FilesToTypesMap } from './introspection/types'
import { getFilesToTypesMap } from './introspection'

export interface NormalizedFile {
  path: string
  defaultName?: string
}

const ajv = new Ajv().addMetaSchema(
  require('ajv/lib/refs/json-schema-draft-06.json'),
)
const validateYaml = ajv.compile(schema)

export function parseConfig(): GraphQLGenDefinition {
  if (!fs.existsSync('graphqlgen.yml')) {
    console.error(chalk.default.red(`No graphqlgen.yml found`))
    process.exit(1)
  }

  const config = yaml.safeLoad(
    fs.readFileSync('graphqlgen.yml', 'utf-8'),
  ) as GraphQLGenDefinition

  if (!validateYaml(config)) {
    console.error(chalk.default.red(`Invalid graphqlgen.yml file`))
    console.error(chalk.default.red(printErrors(validateYaml.errors!)))
    process.exit(1)
  }

  return config
}

function printErrors(errors: any[]): string {
  return errors
    .map(e => {
      const params = Object.keys(e.params)
        .map(key => `${key}: ${e.params[key]}`)
        .join(', ')
      return `${e.dataPath} ${e.message}. ${params}`
    })
    .join('\n')
}

export function parseContext(
  context: string | undefined,
  outputDir: string,
): ContextDefinition | undefined {
  if (!context) {
    return undefined
  }

  const [filePath, interfaceName] = context.split(':')

  return {
    contextPath: getImportPathRelativeToOutput(filePath, outputDir),
    interfaceName,
  }
}

export function parseSchema(schemaPath: string): GraphQLTypes {
  const [filePath, exportName = 'default'] = schemaPath.split(':')

  // We can assume absolute path is cwd prefixed because
  // gg currently only works when run in a directory with the
  // graphqlgen manifest.
  const absoluteFilePath =
    filePath[0] == '/' ? filePath : Path.join(process.cwd(), filePath)

  if (!fs.existsSync(absoluteFilePath)) {
    console.error(
      chalk.default.red(`The schema file ${filePath} does not exist`),
    )
    process.exit(1)
  }

  let schema: string | undefined
  try {
    if (filePath.endsWith('.ts')) {
      tsNode.register({
        transpileOnly: true,
      })
      const schemaModule = require(absoluteFilePath)
      const loadedSchema = schemaModule[exportName]

      if (typeof loadedSchema === 'string') {
        schema = loadedSchema
      } else {
        schema = print(loadedSchema)
      }
    } else {
      schema = importSchema(filePath)
    }
  } catch (e) {
    console.error(
      chalk.default.red(`Error occurred while reading schema: ${e}`),
    )
    process.exit(1)
  }

  let types: GraphQLTypes
  try {
    types = extractTypes(schema!)
  } catch (e) {
    console.error(chalk.default.red(`Failed to parse schema: ${e}`))
    process.exit(1)
  }

  return types!
}

function buildModel(
  modelName: string,
  filePath: string,
  filesToTypesMap: FilesToTypesMap,
  outputDir: string,
  language: Language,
): Model {
  const absoluteFilePath = getAbsoluteFilePath(filePath, language)
  const importPathRelativeToOutput = getImportPathRelativeToOutput(
    absoluteFilePath,
    outputDir,
  )
  return {
    absoluteFilePath,
    importPathRelativeToOutput,
    definition: filesToTypesMap[filePath][modelName],
  }
}

export function getPath(file: File): string {
  if (typeof file === 'string') {
    return file
  }

  return file.path
}

export function getDefaultName(file: File): string | null {
  if (typeof file === 'string') {
    return null
  }

  return file.defaultName || null
}

export function normalizeFiles(
  files: File[] | undefined,
  language: Language,
): NormalizedFile[] {
  return files !== undefined
    ? files.map(file => ({
        defaultName: typeof file === 'object' ? file.defaultName : undefined,
        path: normalizeFilePath(getPath(file), language),
      }))
    : []
}

export function parseModels(
  models: Models,
  schema: GraphQLTypes,
  outputDir: string,
  language: Language,
): ModelMap {
  const graphQLTypes = extractGraphQLTypesWithoutRootsAndInputsAndEnums(schema)
  const normalizedFiles = normalizeFiles(models.files, language)
  const filesToTypesMap = getFilesToTypesMap()
  const overriddenModels = !!models.override ? models.override : {}
  const typeToFileMapping = getTypeToFileMapping(
    normalizedFiles,
    filesToTypesMap,
  )

  return graphQLTypes.reduce((acc, type) => {
    if (overriddenModels[type.name]) {
      const [filePath, modelName] = models.override![type.name].split(':')

      return {
        ...acc,
        [type.name]: buildModel(
          modelName,
          normalizeFilePath(filePath, language),
          filesToTypesMap,
          outputDir,
          language,
        ),
      }
    }

    const typeFileTuple = Object.entries(typeToFileMapping).find(
      ([typeName, file]) => {
        const defaultName = getDefaultName(file)
        const replacedTypeName = defaultName
          ? replaceVariablesInString(defaultName, { typeName: type.name })
          : type.name

        return typeName === replacedTypeName
      },
    )

    if (!typeFileTuple) {
      throw new Error(
        `Could not find type ${type.name} in any of the provided files`,
      )
    }

    const file = typeFileTuple[1]

    const filePath = getPath(file)
    const defaultName = getDefaultName(file)

    const replacedTypeName = defaultName
      ? replaceVariablesInString(defaultName, { typeName: type.name })
      : type.name

    return {
      ...acc,
      [type.name]: buildModel(
        replacedTypeName,
        filePath,
        filesToTypesMap,
        outputDir,
        language,
      ),
    }
  }, {})
}

interface ReplacementMap {
  [key: string]: string
}

export function replaceVariablesInString(
  str: string,
  replacements: ReplacementMap,
) {
  const variableSyntax = RegExp(
    '\\${([ ~:a-zA-Z0-9._\'",\\-\\/\\(\\)]+?)}',
    'g',
  )
  let newStr = str
  if (variableSyntax.test(str)) {
    str.match(variableSyntax)!.forEach(matchedString => {
      // strip ${} away to get the pure variable name
      const variableName = matchedString
        .replace(variableSyntax, (_, varName) => varName.trim())
        .replace(/\s/g, '')

      if (replacements[variableName]) {
        newStr = replaceAll(newStr, matchedString, replacements[variableName])
      } else {
        throw new Error(`Variable ${variableName} is not covered by a value`)
      }
    })
  }

  return newStr
}
