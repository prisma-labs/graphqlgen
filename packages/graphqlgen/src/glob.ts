import * as glob from 'glob'
import { File } from 'graphqlgen-json-schema'
import { getPath } from './parse'

/**
 * Returns the path array from glob patterns
 */
export const extractGlobPattern = (file: File) => {
  return glob.sync(getPath(file))
}

/**
 * Handles the glob pattern of models.files
 */
export const handleGlobPattern = (files?: File[]): File[] => {
  if (!files) {
    return []
  }

  return files.reduce<File[]>((acc, file) => {
    const globedPaths = extractGlobPattern(file)

    if (globedPaths.length === 0) {
      return [...acc, file]
    }

    const globedFiles: File[] = globedPaths.map(path => {
      if (typeof file === 'string') {
        return path
      }

      return {
        path,
        defaultName: file.defaultName,
      }
    })

    return [...acc, ...globedFiles]
  }, [])
}
