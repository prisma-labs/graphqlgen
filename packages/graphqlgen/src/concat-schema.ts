import chalk from 'chalk'
import * as fs from 'fs'
import { GraphQLGenDefinition } from 'graphqlgen-json-schema'
import * as mkdirp from 'mkdirp'
import * as path from 'path'

import { extractGlobPattern } from './glob'

export const readSchemaSync = (schemaDir: string) => {
  try {
    const schemas = extractGlobPattern([schemaDir])

    if (schemas) {
      return schemas
        .map(f => {
          return fs.readFileSync(f).toString()
        })
        .join('\n')
    }
  } catch (error) {
    console.log(error)
  }
}

export const generateSchema = (config: GraphQLGenDefinition) => {
  const data = readSchemaSync(config.schema)

  const outputPath = path.join(process.cwd(), config['schema-output'])

  // Create generation target folder, if it does not exist
  mkdirp.sync(path.dirname(outputPath))

  try {
    console.log(chalk.cyan('Generating new schema :', outputPath))
    fs.writeFileSync(outputPath, data, { flag: 'w' })
    return true
  } catch (error) {
    console.log(error)
    return error
  }
}
