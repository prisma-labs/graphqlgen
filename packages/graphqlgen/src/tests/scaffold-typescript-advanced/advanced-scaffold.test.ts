import {
  createConfig,
  defaultConfig,
} from 'graphqlgen-json-schema/dist/default-config'
import { readSchemaSync, generateSchema } from '../../concat-schema'
import { parseSchema, parseModels } from '../../parse'
import { generateCode } from '../..'
import { extractGlobPattern, handleGlobPattern } from '../../glob'

const config = createConfig('./src/tests/fixtures/advanced-scaffold')

test('default config', async () => {
  expect(defaultConfig).toMatchSnapshot()
})

test('create config', async () => {
  expect(config).toMatchSnapshot()
})

test('Check Schemas', () => {
  expect(extractGlobPattern([config.schema])).toMatchSnapshot()
})

test('Check Models', () => {
  expect(extractGlobPattern(config.models.files as any)).toMatchSnapshot()
})

test('Read Schema', () => {
  expect(readSchemaSync(config.schema)).toMatchSnapshot()
})

test('Generate Schema', () => {
  expect(generateSchema(config)).toBeTruthy()
})

test('Advance scaffolding', () => {
  const language = config.language

  generateSchema(config)
  config.schema = config['schema-output']

  // Override the config.models.files using handleGlobPattern
  config.models = {
    ...config.models,
    files: handleGlobPattern(config.models.files),
  }

  const schema = parseSchema(config.schema)

  const modelMap = parseModels(config.models, schema, config.output, language)
  const { generatedTypes, generatedResolvers } = generateCode({
    schema,
    language,
    config,
    modelMap,
    prettify: true,
  })
  expect(generatedTypes).toMatchSnapshot()
  expect(generatedResolvers).toMatchSnapshot()
})
