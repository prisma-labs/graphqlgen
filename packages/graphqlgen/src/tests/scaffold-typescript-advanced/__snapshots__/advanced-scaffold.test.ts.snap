// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Advance scaffolding 1`] = `
"// Code generated by github.com/prisma/graphqlgen, DO NOT EDIT.

import { GraphQLResolveInfo } from \\"graphql\\";
import { Context } from \\"../models/context\\";
import { Schema1 } from \\"../models/definitions\\";
import { Schema2 } from \\"../models/definitions\\";
import { Schema3 } from \\"../models/definitions\\";

export namespace Schema1Resolvers {
  export const defaultResolvers = {
    schema1: (parent: Schema1) => parent.schema1
  };

  export type Schema1Resolver = (
    parent: Schema1,
    args: {},
    ctx: Context,
    info: GraphQLResolveInfo
  ) => string | Promise<string>;

  export interface Type {
    schema1: (
      parent: Schema1,
      args: {},
      ctx: Context,
      info: GraphQLResolveInfo
    ) => string | Promise<string>;
  }
}

export namespace Schema2Resolvers {
  export const defaultResolvers = {
    schema2: (parent: Schema2) => parent.schema2
  };

  export type Schema2Resolver = (
    parent: Schema2,
    args: {},
    ctx: Context,
    info: GraphQLResolveInfo
  ) => string | Promise<string>;

  export interface Type {
    schema2: (
      parent: Schema2,
      args: {},
      ctx: Context,
      info: GraphQLResolveInfo
    ) => string | Promise<string>;
  }
}

export namespace Schema3Resolvers {
  export const defaultResolvers = {
    schema3: (parent: Schema3) => parent.schema3
  };

  export type Schema3Resolver = (
    parent: Schema3,
    args: {},
    ctx: Context,
    info: GraphQLResolveInfo
  ) => string | Promise<string>;

  export interface Type {
    schema3: (
      parent: Schema3,
      args: {},
      ctx: Context,
      info: GraphQLResolveInfo
    ) => string | Promise<string>;
  }
}

export interface Resolvers {
  Schema1: Schema1Resolvers.Type;
  Schema2: Schema2Resolvers.Type;
  Schema3: Schema3Resolvers.Type;
}
"
`;

exports[`Advance scaffolding 2`] = `
Array [
  Object {
    "code": "// This resolver file was scaffolded by github.com/prisma/graphqlgen, DO NOT EDIT.
// Please do not import this file directly but copy & paste to your application code.

import { Schema1Resolvers } from \\"[TEMPLATE-INTERFACES-PATH]\\";

export const Schema1: Schema1Resolvers.Type = {
  ...Schema1Resolvers.defaultResolvers
};
",
    "force": false,
    "path": "Schema1.ts",
  },
  Object {
    "code": "// This resolver file was scaffolded by github.com/prisma/graphqlgen, DO NOT EDIT.
// Please do not import this file directly but copy & paste to your application code.

import { Schema2Resolvers } from \\"[TEMPLATE-INTERFACES-PATH]\\";

export const Schema2: Schema2Resolvers.Type = {
  ...Schema2Resolvers.defaultResolvers
};
",
    "force": false,
    "path": "Schema2.ts",
  },
  Object {
    "code": "// This resolver file was scaffolded by github.com/prisma/graphqlgen, DO NOT EDIT.
// Please do not import this file directly but copy & paste to your application code.

import { Schema3Resolvers } from \\"[TEMPLATE-INTERFACES-PATH]\\";

export const Schema3: Schema3Resolvers.Type = {
  ...Schema3Resolvers.defaultResolvers
};
",
    "force": false,
    "path": "Schema3.ts",
  },
  Object {
    "code": "// This resolver file was scaffolded by github.com/prisma/graphqlgen, DO NOT EDIT.
// Please do not import this file directly but copy & paste to your application code.

import { Resolvers } from \\"[TEMPLATE-INTERFACES-PATH]\\";

import { Schema1 } from \\"./Schema1\\";
import { Schema2 } from \\"./Schema2\\";
import { Schema3 } from \\"./Schema3\\";

export const resolvers: Resolvers = {
  Schema1,
  Schema2,
  Schema3
};
",
    "force": false,
    "path": "index.ts",
  },
]
`;

exports[`Check Models 1`] = `
Array [
  "./src/tests/fixtures/advanced-scaffold/models/context.ts",
  "./src/tests/fixtures/advanced-scaffold/models/definitions.ts",
]
`;

exports[`Check Schemas 1`] = `
Array [
  "./src/tests/fixtures/advanced-scaffold/schemas/schema-1.graphql",
  "./src/tests/fixtures/advanced-scaffold/schemas/schema-2.graphql",
  "./src/tests/fixtures/advanced-scaffold/schemas/schema-3.graphql",
]
`;

exports[`Read Schema 1`] = `
"type Schema1 {
  schema1: String!
}

type Schema2 {
  schema2: String!
}

type Schema3 {
  schema3: String!
}
"
`;

exports[`create config 1`] = `
Object {
  "context": "./src/tests/fixtures/advanced-scaffold/models/context.ts:Context",
  "language": "typescript",
  "models": Object {
    "files": Array [
      "./src/tests/fixtures/advanced-scaffold/models/**.ts",
    ],
  },
  "output": "./src/tests/fixtures/advanced-scaffold/__generated__/graphqlgen.ts",
  "resolver-scaffolding": Object {
    "layout": "file-per-type",
    "output": "./src/tests/fixtures/advanced-scaffold/__generated__/tmp-resolvers",
  },
  "schema": "./src/tests/fixtures/advanced-scaffold/schemas/**.graphql",
  "schema-output": "./src/tests/fixtures/advanced-scaffold/__generated__/graphqlgen.schema.graphql",
}
`;
